staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go

```diff
@@ -1070,31 +1066,33 @@ func (vs *VSphere) DisksAreAttached(nodeVolumes map[k8stypes.NodeName][]string)
 				VC_DC := nodeInfo.vcServer + nodeInfo.dataCenter.String()
 				dcNodes[VC_DC] = append(dcNodes[VC_DC], nodeName)
 			}
-
+		
+			rNodes := make(chan []k8stypes.NodeName)
+			ech := make(chan error)
 			for _, nodeNames := range dcNodes {
 				localAttachedMap := make(map[string]map[string]bool)
 				localAttachedMaps = append(localAttachedMaps, localAttachedMap)
 				// Start go routines per VC-DC to check disks are attached
-				go func(nodes []k8stypes.NodeName) {
+				go func(nodes []k8stypes.NodeName, rNodes chan []k8stypes.NodeName, ech chan error) {
 					nodesToRetryLocal, err := vs.checkDiskAttached(ctx, nodes, nodeVolumes, localAttachedMap, retry)
 					if err != nil {
 						if !vclib.IsManagedObjectNotFoundError(err) {
-							globalErrMutex.Lock()
-							globalErr = err
-							globalErrMutex.Unlock()
 							klog.Errorf("Failed to check disk attached for nodes: %+v. err: %+v", nodes, err)
+							ech <- err
 						}
 					}
-					nodesToRetryMutex.Lock()
-					nodesToRetry = append(nodesToRetry, nodesToRetryLocal...)
-					nodesToRetryMutex.Unlock()
-					wg.Done()
-				}(nodeNames)
-				wg.Add(1)
+					rNodes <- nodesToRetryLocal
```

Comment:
Note that channels are usually 2x-10x slower than locks in scenarios where you are protecting access to a resource.  Channels are almost as fast if you have to copy resources under the lock.

Reply:
@smarterclayton , thanks for the valuable insight. I was actually doing this as an experiment to better understand the performance difference(s). I wrote a small program to test this using 10000 goroutines using a lock and a buffered channel to synchronize write(s) to a global slice. The locks are fasterr than the channels by  4-5x.
https://play.golang.com/p/dBJoTwo2q05

Locks			Channels
5.412399ms		24.385236ms
7.19043ms		26.541433ms
6.987753ms		24.734029ms
6.842185ms		22.605961ms
6.377317ms		29.177046ms

cc @misterikkit 

My system architechture is: 
Linux linux 5.3.0-46-generic #38~18.04.1-Ubuntu SMP Tue Mar 31 04:17:56 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux